_posts/2017-11-21-javascript-modules.md

```diff
@@ -0,0 +1,214 @@
+---
+layout: post
+title: JavaScript Modules
+date: '2017-11-21'
+---
+
+Today, I looked a bit closer at JavaScript Modules. I have been using modules since I started using JavaScript but I must admit I have never looked into the subject deeply. Using [Babel](https://babeljs.io/) (a JavaScript transpiler) and [webpack](https://webpack.js.org/) (a JavaScript bundler) things have been working fine and I just assumed webpack built a sort of dependency tree and Babel helped it understand the latest ES module syntax (import/export).
+<br/>
+This post is about what I learnt and understood about modules in the context of my setup (babel + webpack). To be honest this is from me to me, if you want to learn about modules and you happen to read this post, be critical.
+
+#### JavaScript Modules
+
+First of all, what is a module? The idea behind module is to isolate a piece of code so that it is easier to maintain, remove, modify, use. The concept of modules exist in many languages but might be called different things: C# namespace, Java package, python modules, etc...
+
+In the JavaScript language, modules have only been available natively since 2015 (wow!) with the ECMAScript 6 [standard](http://www.ecma-international.org/ecma-262/6.0/ECMA-262.pdf). Not being able to break down code into modules is madness (in my opinion) and the community has come up with other ways to add modules to the language. (I am not really sure why modules were not added before, I don't know what the process is to create a new ECMAScript standard).
+<br/>
+The most famous module standards (pre ES6) are [AMD](https://github.com/amdjs/amdjs-api/blob/master/AMD.md)
+(a browser targeted standard whose most famous implementation is [requireJS](http://requirejs.org/)) and [CommonJS Modules](http://wiki.commonjs.org/wiki/Modules/1.1) (a standard defined by the group [CommonJS](https://en.wikipedia.org/wiki/CommonJS)).
+
+So depending on the context, people may refer to ES modules (ESM), CommonJS (CJS) modules or AMD (I don't think there is even consistency in acronym 🤔).
```

Comment:
I still 🤔 this needs to be "acronyms"rather than "acronym" :)

Reply:
oops